### PR TITLE
IGNITE-14477: added ducktape rebalance persistent_test.

### PR DESCRIPTION
**Pull request by Sega76**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py [hunk 1]
```diff
@@ -287,17 +288,18 @@ def event_time(evt_message, node):
 
         return datetime.strptime(match.group(), "[%Y-%m-%d %H:%M:%S,%f]") if match else None
 
-    def get_event_time_on_node(self, node, log_pattern, from_the_beginning=True, timeout=15):
+    @staticmethod
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py:291]: What a value this change brings? A suppose that not at all.

File: modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py [hunk 2]
```diff
@@ -16,17 +16,43 @@
...
 # pylint: disable=too-many-arguments
-def preload_data(context, config, preloaders, backups, cache_count, entry_count, entry_size, timeout=3600):
+from ignitetest.services.utils.ignite_aware import IgniteAwareService
+from ignitetest.services.utils.ignite_configuration import IgniteConfiguration, DataStorageConfiguration
+from ignitetest.services.utils.ignite_configuration.data_storage import DataRegionConfiguration
+from ignitetest.utils.enum import constructible
+from ignitetest.utils.version import IgniteVersion
+
+NUM_NODES = 4
+DEFAULT_DATA_REGION_SZ = 1024 * 1024 * 1024
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py:38]: `1 << 30`

File: modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py [hunk 3]
```diff
@@ -16,17 +16,43 @@
 """
 This package contains rebalance tests.
 """
-
+import sys
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py:19]: Why did you put test in `__init__.py` It is a wrong place for it

**Review by ivandasch — CHANGES_REQUESTED** [review]

File: modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py (hunk 3 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py]: This is not a test, general functions for rebalancing tests. transferred this to the rebalance/util.py

File: modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py (hunk 2 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/tests/rebalance/__init__.py]: done

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py (hunk 1 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py]: fixed

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py [hunk 4]
```diff
@@ -548,6 +549,41 @@ def restore_from_snapshot(self, snapshot_name: str):
...
+        metric_exporter='org.apache.ignite.spi.metric.jmx.JmxMetricExporterSpi'
+        to the config.
+
+        :param node: Node.
+        :param timeout_sec: Timeout to wait the rebalance to complete.
+        """
+
+        delta_time = datetime.now() + timedelta(seconds=timeout_sec)
+
+        _node = node if node else random.choice(self.alive_nodes)
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py:565]: It is ok just to write
```
if not node:
    node = random.choice(self.alive_nodes)

```

File: modules/ducktests/tests/ignitetest/tests/rebalance/util.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+            "end_time": {},
+            "duration": {},
+            "received_bytes": {}
+        }
+
+        metrics = list(map(lambda node: get_rebalance_metrics(node, cache_name), nodes))
+
+        def __key(tup):
+            return tup[1]
+
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/rebalance/util.py:179]: Lets rewrite it in loop:
```
from itertools import product

for prop, mode in list(product(['start_time', 'end_time'],['min', 'max'])) + list(product(['duration','received_bytes'], ['min', 'max', 'sum'])):
      if mode == 'sum':
            val = sum(map(lambda item: getattr(item, prop)))
      else:
            val = map(lambda item: (item.node, getitem(item, prop)), metrics
            val = min(val, key=__key) if mode == 'min' else max(val, key=__key)
      stats[prop][mode] = val


```

File: modules/ducktests/tests/ignitetest/tests/rebalance/util.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    """
+    assert preloaders > 0
+    assert cache_count > 0
+    assert entry_count > 0
+    assert entry_size > 0
+
+    apps = []
+
+    def start_app(from_, to_):
+        app0 = IgniteApplicationService(
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/rebalance/util.py:70]: Why do you use app0, app1?

File: modules/ducktests/tests/ignitetest/tests/rebalance/util.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+                "from": from_,
+                "to": to_
+            },
+            shutdown_timeout_sec=timeout)
+        app0.start_async()
+
+        apps.append(app0)
+
+    count = int(entry_count / preloaders)
+    _from = 0
```
Comment (ivandasch) [inline modules/ducktests/tests/ignitetest/tests/rebalance/util.py:87]: IMHO, it is weid naming. May be just start:end? Or begin:end?

File: modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py (hunk 4 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/services/utils/ignite_aware.py]: fixed

File: modules/ducktests/tests/ignitetest/tests/rebalance/util.py (hunk 6 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/tests/rebalance/util.py]: it was, I just migrated to util.py
fixed

File: modules/ducktests/tests/ignitetest/tests/rebalance/util.py (hunk 7 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/tests/rebalance/util.py]: fixed

File: modules/ducktests/tests/ignitetest/tests/rebalance/util.py (hunk 5 above)
Comment (Sega76) [reply modules/ducktests/tests/ignitetest/tests/rebalance/util.py]: done, thanks

**Review by ivandasch — APPROVED** [review]

**Pull request merged**